test/unit/org/apache/cassandra/index/sai/cql/IndexQuerySupport.java

```diff
@@ -0,0 +1,486 @@
+
+                andQuery(tester, model,
+                         DataModel.TEXT_COLUMN, Operator.EQ, "Alaska",
+                         firstPartitionKey, Operator.EQ, 0,
+                         secondPrimaryKey, Operator.GTE, -1);
+
+                andQuery(tester, model,
+                         DataModel.TEXT_COLUMN, Operator.EQ, "Kentucky",
+                         firstPartitionKey, Operator.GT, 4,
+                         secondPrimaryKey, Operator.LT, 50);
+
+                andQuery(tester, model,
+                         DataModel.TEXT_COLUMN, Operator.EQ, "Wyoming",
+                         firstPartitionKey, Operator.LT, 200,
+                         secondPrimaryKey, Operator.GT, 0);
+            }
+        }
+
+        void query(DataModel.Executor tester, DataModel model, String column, Operator operator, Object value) throws Throwable
+        {
+            String query = String.format(DataModel.SIMPLE_SELECT_TEMPLATE, DataModel.ASCII_COLUMN, column, operator);
+            String queryValidator = String.format(DataModel.SIMPLE_SELECT_WITH_FILTERING_TEMPLATE, DataModel.ASCII_COLUMN, column, operator);
+            validate(tester, model, query, queryValidator, value, limit);
+        }
+
+        void andQuery(DataModel.Executor tester, DataModel model,
+                      String column1, Operator operator1, Object value1,
+                      String column2, Operator operator2, Object value2,
+                      boolean filtering) throws Throwable
+        {
+            String query = String.format(filtering ? DataModel.TWO_CLAUSE_AND_QUERY_FILTERING_TEMPLATE : DataModel.TWO_CLAUSE_AND_QUERY_TEMPLATE,
+                                         DataModel.ASCII_COLUMN, column1, operator1, column2, operator2);
+
+            String queryValidator = String.format(DataModel.TWO_CLAUSE_AND_QUERY_FILTERING_TEMPLATE,
+                                                  DataModel.ASCII_COLUMN, column1, operator1, column2, operator2);
+
+            validate(tester, model,query, queryValidator, value1, value2, limit);
+        }
+
+        void andQuery(DataModel.Executor tester, DataModel model,
+                      String column1, Operator operator1, Object value1,
+                      String column2, Operator operator2, Object value2,
+                      String column3, Operator operator3, Object value3) throws Throwable
+        {
+            // TODO: If we support indexes in all columns, ALLOW FILTERING might go away here...
+            String query = String.format(DataModel.THREE_CLAUSE_AND_QUERY_FILTERING_TEMPLATE,
+                                         DataModel.ASCII_COLUMN, column1, operator1, column2, operator2, column3, operator3);
+
+            String queryValidator = String.format(DataModel.THREE_CLAUSE_AND_QUERY_FILTERING_TEMPLATE,
+                                                  DataModel.ASCII_COLUMN, column1, operator1, column2, operator2, column3, operator3);
```

Comment:
This is correct but in this case I would prefer to leave them both here. The `query` is run against the indexed table and the `queryValidator` is run against the non-indexed table. At present they both need `ALLOW FILTERING` because the `StatementRestriction` changes are not in place to allow multiple index queries without `ALLOW FILTERING`.

Reply:
Ah, I see. Some day we'll have a version of the quey without AF. Looks fine to me.

Reply:
@mike-tr-adamson Which Jira is likely to contain the `StatementRestriction` changes?